### PR TITLE
Feature/applied block counting optimize fr

### PR DESCRIPTION
**Pull request by eu321**

### CODE REVIEW

File: include/veriblock/blockchain/pop/pop_state_machine.hpp [hunk 1]
```diff
@@ -202,6 +203,7 @@ struct PopStateMachine {
     }
 
     index.unsetFlag(BLOCK_APPLIED);
+    --ed_.appliedBlockCount;
```
Comment (blackyblack) [inline include/veriblock/blockchain/pop/pop_state_machine.hpp:206]: I'd add an assert that appliedBlockCount is not zero.

Comment (eu321) [reply include/veriblock/blockchain/pop/pop_state_machine.hpp:206]: there's a more elaborate check in setState() that would also catch this case, but it doesn't hurt, yeah

File: include/veriblock/blockchain/base_block_tree.hpp [hunk 2]
```diff
@@ -340,6 +342,14 @@ struct BaseBlockTree {
     return validity_sig_.disconnect(id);
   }
 
+  index_t& getRoot() {
```
Comment (Warchant) [inline include/veriblock/blockchain/base_block_tree.hpp:345]: In our terms this is called "bootstrap block" 
```suggestion
  index_t& getBootstrapBlock() {
```

File: include/veriblock/blockchain/base_block_tree.hpp [hunk 3]
```diff
@@ -340,6 +342,14 @@ struct BaseBlockTree {
...
+    VBK_ASSERT_MSG(isBootstrapped(), "must be bootstrapped");
+    return *getBestChain().first();
+  }
+
+  //! the number of blocks that have BLOCK_APPLIED flag set
+  block_height_t appliedBlockCount = 0;
```
Comment (Warchant) [inline include/veriblock/blockchain/base_block_tree.hpp:351]: should it be public?

File: include/veriblock/blockchain/pop/fork_resolution.hpp [hunk 4]
```diff
@@ -506,7 +510,7 @@ struct PopAwareForkResolutionComparator {
     {
       auto guard = ing_->deferForkResolutionGuard();
 
-      if (!sm.apply(*chainB.first(), *chainB.tip(), state, false)) {
+      if (!sm.apply(*chainB.first(), *chainB.tip(), state)) {
```
Comment (Warchant) [inline include/veriblock/blockchain/pop/fork_resolution.hpp:513]: why did you change this? 
this apply applies chain B together with chain A, and this `false` indicated that we should not set flag `CAN_BE_APPLIED` during this state change

File: include/veriblock/blockchain/pop/pop_state_machine.hpp [hunk 5]
```diff
@@ -170,10 +163,18 @@ struct PopStateMachine {
                  "its payloads");
     }
 
-    index.setFlag(BLOCK_APPLIED);
-    if (shouldSetCanBeApplied) {
+    // if the applied block count equals the size of the chain between the root
```
Comment (Warchant) [inline include/veriblock/blockchain/pop/pop_state_machine.hpp:166]:
```suggestion
    // if the applied block count equals the size of the chain between the bootstrap block
```

**Comment by Warchant** [comment]
Please open same PR based on `release/beta` branch

File: include/veriblock/blockchain/base_block_tree.hpp (hunk 3 above)
Comment (eu321) [reply include/veriblock/blockchain/base_block_tree.hpp:351]: either it has to be public since it's accessed by other classes, or we need to add blockApplied() blockUnapplied(), getAppliedBlockCount()

File: include/veriblock/blockchain/pop/pop_state_machine.hpp (hunk 5 above)
Comment (eu321) [reply include/veriblock/blockchain/pop/pop_state_machine.hpp:166]: there can be several bootstrap blocks, while the root block refers to the root of the tree

Comment (Warchant) [reply include/veriblock/blockchain/pop/pop_state_machine.hpp:166]: ok, add a comment `the root block (first bootstrap block)`

**Review by Warchant — APPROVED** [review]
